File: leetcode/C_04_stack/D2024_01_13_do_8_Q394_decode_string/solution.go
package main

import (
	"bytes"
	"fmt"
	"unicode"
)

func decodeString(s string) string {
	var buffer bytes.Buffer
	n := 0
	stack_number := []int{}
	stack_str := []string{}

	for _, item := range s {
		item_char := string(item)
		fmt.Println("item", item_char)

		if unicode.IsDigit(item_char) {
			n = n*10 + (item_char - "0")
		} else if item_char == "[" {
			append(stack_number, n)
			n = 0
			append(stack_str, buffer)
			buffer = *new(bytes.Buffer)
		} else if item_char == "]" {
			k := stack_number[len(stack_number)-1]
			stack_number = stack_number[:len(stack_number)-1]
			var tmp bytes.Buffer = buffer
			buffer = stack_str[len(stack_str)-1]
			stack_str = stack_str[:len(stack_str)-1]
			for ; k > 0; k-- {
				buffer.WriteString(tmp.String())
			}
		} else {
			buffer.WriteString(item_char)
		}
	}

	return buffer.String()
}

func main() {
	decodeString("abc")
	fmt.Println("run main")
}

File: leetcode/C_04_stack/D2024_01_13_do_8_Q394_decode_string/solution.ts
function isDigit(char: string): boolean {
  return /^\d$/.test(char);
}

// Usage
console.log(isDigit("5")); // true
console.log(isDigit("a")); // false
console.log(isDigit(" ")); // false

function decodeString(s: string): string {
  let stringBuilder: string = "";
  let n: number = 0;
  const stack_nubmer: number[] = [];
  const stack_str: string[] = [];

  for (const c of s) {
    console.log("c", c);
    if (isDigit(c)) {
      n = n * 10 + parseInt(c);
    } else if (c === "[") {
      stack_nubmer.push(n);
      n = 0;
      stack_str.push(stringBuilder);
      stringBuilder = "";
    } else if (c === "]") {
      let k = stack_nubmer.pop() ?? 0;
      const tmp = stringBuilder;
      stringBuilder = stack_str.pop() ?? "";
      while (k > 0) {
        stringBuilder += tmp;
        k--;
      }
    } else {
      stringBuilder += c;
    }
  }

  return stringBuilder;
}

decodeString("3[a]2[bc]"); // "aaabcbc"
